verifier/src/report_body.rs

```diff
@@ -0,0 +1,164 @@
+// Copyright (c) 2018-2023 The MobileCoin Foundation
+
+//! Verifiers which operate on the [`ReportBody`]
+
+use crate::{VerificationError, Verifier};
+use mc_sgx_core_types::{Attributes, ReportBody};
+use subtle::CtOption;
+
+/// Verify the report body is as expected.
+#[derive(Clone, Debug, Eq, Hash, PartialEq)]
+pub struct ReportBodyVerifier<'a> {
+    expected_report_body: ReportBody,
+    report_body: &'a ReportBody,
+}
+
+impl<'a> ReportBodyVerifier<'a> {
+    /// Create a new [`ReportBodyVerifier`] instance.
+    ///
+    /// # Arguments:
+    /// * `expected_report_body` - The expected report body.
+    /// * `report_body` - The report body to verify conforms to the
+    ///   `expected_report_body`.
+    pub fn new(expected_report_body: ReportBody, report_body: &'a ReportBody) -> Self {
+        Self {
+            expected_report_body,
+            report_body,
+        }
+    }
+}
+
+impl Verifier for ReportBodyVerifier<'_> {
+    type Error = VerificationError;
+
+    fn verify(&self) -> CtOption<Self::Error> {
+        AttributesVerifier::new(self.expected_report_body.attributes(), self.report_body).verify()
+    }
+}
+
+/// Verify the attributes are as expected.
+#[derive(Clone, Debug, Eq, Hash, PartialEq)]
+pub struct AttributesVerifier<'a> {
+    expected_attributes: Attributes,
+    report_body: &'a ReportBody,
+}
+
+impl<'a> AttributesVerifier<'a> {
+    /// Create a new [`AttributesVerifier`] instance.
+    ///
+    /// # Arguments:
+    /// * `expected_attributes` - The expected attributes.
+    /// * `report_body` - The report body containing attributes that conforms
+    ///    to the `expected_attributes`.
+    pub fn new(expected_attributes: Attributes, report_body: &'a ReportBody) -> Self {
+        Self {
+            expected_attributes,
+            report_body,
+        }
+    }
+}
+
+impl Verifier for AttributesVerifier<'_> {
+    type Error = VerificationError;
+    fn verify(&self) -> CtOption<Self::Error> {
+        let expected = self.expected_attributes;
+        let actual = self.report_body.attributes();
+        // TODO - This should be a constant time comparison.
+        let is_some = if expected == actual { 0 } else { 1 };
```

Comment:
I'm up for suggestions here, it may be that in order to implement this I need to flush out how the verifiers are actually getting passed the data they're verifying.
it may be that the verifier is a newtype wrapper around the `Attributes` and then it can implement `ConstantTimeEq`

Reply:
I'd say we should just `impl ConstantTimeEq for Attributes` in `core-types`, honestly.

Reply:
created https://github.com/mobilecoinfoundation/sgx/issues/280